File: configurator_typescript/src/configuration.ts
import { defineScope } from "./scope.js";
import type {
  ConfigImage,
  ConfigScope,
  ConfigTextStyle,
  ConfigTranslations,
} from "./scope.js";

/** Runtime key categories emitted by configuration access subscriptions. */
export const KeyType = Object.freeze({
  Flag: "flag",
  Color: "color",
  Image: "image",
  ImageList: "imageList",
  Route: "route",
  Size: "size",
  Padding: "padding",
  Margin: "margin",
  Misc: "misc",
  TextStyle: "textStyle",
  // Keep the descriptive alias while serializing the event exactly as Dart.
  Translation: "string",
  String: "string",
} as const);

export type KeyType = (typeof KeyType)[keyof typeof KeyType];

interface AccessEvent<TType extends KeyType, TKey, TValue> {
  readonly type: TType;
  readonly scope: ConfigScope;
  readonly key: TKey;
  readonly value: TValue;
}

/** A discriminated event emitted after a configuration value is found. */
export type ConfigAccessEvent =
  | AccessEvent<typeof KeyType.Flag, string, boolean>
  | AccessEvent<typeof KeyType.Color, string, string>
  | AccessEvent<typeof KeyType.Image, string, string>
  | AccessEvent<typeof KeyType.ImageList, string, readonly string[]>
  | AccessEvent<typeof KeyType.Route, number, string>
  | AccessEvent<typeof KeyType.Size, string, number>
  | AccessEvent<typeof KeyType.Padding, string, number>
  | AccessEvent<typeof KeyType.Margin, string, number>
  | AccessEvent<typeof KeyType.Misc, string, unknown>
  | AccessEvent<typeof KeyType.TextStyle, string, ConfigTextStyle>
  | AccessEvent<typeof KeyType.Translation, string, ConfigTranslations>;

/** Dart-compatible name for configuration access events. */
export type ConfigKeyLog = ConfigAccessEvent;

export type Unsubscribe = () => void;
export type ConfigurationListener = (configuration: Configuration) => void;
export type AccessListener = (event: ConfigAccessEvent) => void;
export type ScopePredicate = (scope: ConfigScope) => boolean;

export interface ConfigurationOptions {
  readonly scopes?: readonly ConfigScope[];
}

export interface MutationOptions {
  readonly notify?: boolean;
}

export interface PushScopeOptions extends MutationOptions {
  /** Skip a push when the exact same scope is already current. */
  readonly checkEquality?: boolean;
}

interface LookupResult<T> {
  readonly scope: ConfigScope;
  readonly value: T;
}

const EMPTY_TEXT_STYLE: ConfigTextStyle = Object.freeze({});
const EMPTY_TRANSLATIONS: ConfigTranslations = Object.freeze({});
const EMPTY_IMAGE_LIST: readonly string[] = Object.freeze([]);

let defaultScopeSequence = 0;

function createDefaultScope(): ConfigScope {
  defaultScopeSequence += 1;
  return defineScope({
    name: `default-${Date.now()}-${defaultScopeSequence}`,
  });
}

function isScopeArray(
  value: readonly ConfigScope[] | ConfigurationOptions,
): value is readonly ConfigScope[] {
  return Array.isArray(value);
}

/**
 * Resolves generated configuration scopes using weight and stack order.
 *
 * Higher weights win. When two matching scopes have the same weight, the
 * scope inserted later wins.
 */
export class Configuration {
  private scopesValue: ConfigScope[];
  private readonly changeListeners = new Set<ConfigurationListener>();
  private readonly accessListeners = new Set<AccessListener>();

  public constructor(
    scopesOrOptions: readonly ConfigScope[] | ConfigurationOptions = [],
  ) {
    const scopes = isScopeArray(scopesOrOptions)
      ? scopesOrOptions
      : (scopesOrOptions.scopes ?? []);

    this.scopesValue = scopes.length > 0 ? [...scopes] : [createDefaultScope()];
  }

  /** A detached snapshot in insertion order. */
  public get scopes(): readonly ConfigScope[] {
    return [...this.scopesValue];
  }

  /** The most recently inserted scope, independent of its weight. */
  public get currentScopeName(): string {
    return this.scopesValue[this.scopesValue.length - 1]!.name;
  }

  /** Subscribe to scope-stack changes. The callback is not invoked immediately. */
  public subscribe(listener: ConfigurationListener): Unsubscribe {
    this.changeListeners.add(listener);
    return () => {
      this.changeListeners.delete(listener);
    };
  }

  /** Subscribe to successful key lookups. Missing keys do not emit events. */
  public subscribeAccess(listener: AccessListener): Unsubscribe {
    this.accessListeners.add(listener);
    return () => {
      this.accessListeners.delete(listener);
    };
  }

  /** Push a scope and return whether the stack changed. */
  public pushScope(
    scope: ConfigScope,
    options: PushScopeOptions = {},
  ): boolean {
    const checkEquality = options.checkEquality ?? true;
    const current = this.scopesValue[this.scopesValue.length - 1];

    if (checkEquality && current === scope) {
      return false;
    }

    this.scopesValue.push(scope);
    this.emitChange(options.notify ?? true);
    return true;
  }

  /** Pop the current scope without ever removing the final fallback scope. */
  public popScope(options: MutationOptions = {}): ConfigScope | undefined {
    if (this.scopesValue.length <= 1) {
      return undefined;
    }

    const removed = this.scopesValue.pop();
    this.emitChange(options.notify ?? true);
    return removed;
  }

  /** Pop scopes until the current scope satisfies `predicate`. */
  public popScopeUntil(
    predicate: ScopePredicate,
    options: MutationOptions = {},
  ): readonly ConfigScope[] {
    const removed: ConfigScope[] = [];

    while (
      this.scopesValue.length > 1
      && !predicate(this.scopesValue[this.scopesValue.length - 1]!)
    ) {
      removed.push(this.scopesValue.pop()!);
    }

    if (removed.length > 0) {
      this.emitChange(options.notify ?? true);
    }

    return removed;
  }

  /** Remove every matching scope, installing a fresh fallback if none remain. */
  public removeScopeWhere(
    predicate: ScopePredicate,
    options: MutationOptions = {},
  ): readonly ConfigScope[] {
    const kept: ConfigScope[] = [];
    const removed: ConfigScope[] = [];

    for (const scope of this.scopesValue) {
      (predicate(scope) ? removed : kept).push(scope);
    }

    if (removed.length === 0) {
      return removed;
    }

    this.scopesValue = kept.length > 0 ? kept : [createDefaultScope()];
    this.emitChange(options.notify ?? true);
    return removed;
  }

  /** Remove the most recently inserted matching scope. */
  public removeLastScopeWhere(
    predicate: ScopePredicate,
    options: MutationOptions = {},
  ): ConfigScope | undefined {
    for (let index = this.scopesValue.length - 1; index >= 0; index -= 1) {
      const scope = this.scopesValue[index]!;
      if (!predicate(scope)) {
        continue;
      }

      this.scopesValue.splice(index, 1);
      if (this.scopesValue.length === 0) {
        this.scopesValue.push(createDefaultScope());
      }
      this.emitChange(options.notify ?? true);
      return scope;
    }

    return undefined;
  }

  /** Remove the last scope with a matching name or object identity. */
  public removeScope(
    scopeOrName: ConfigScope | string,
    options: MutationOptions = {},
  ): ConfigScope | undefined {
    return this.removeLastScopeWhere(
      typeof scopeOrName === "string"
        ? (scope) => scope.name === scopeOrName
        : (scope) => scope === scopeOrName,
      options,
    );
  }

  public flag(key: string): boolean {
    const result = this.findRecordValue((scope) => scope.flags, key);
    if (result === undefined) {
      return false;
    }

    this.emitAccess({ type: KeyType.Flag, ...result, key });
    return result.value;
  }

  public color(key: string): string {
    const result = this.findRecordValue((scope) => scope.colors, key);
    if (result === undefined) {
      return "";
    }

    this.emitAccess({ type: KeyType.Color, ...result, key });
    return result.value;
  }

  public image(key: string): string {
    const result = this.findRecordValue((scope) => scope.images, key);
    if (result === undefined || typeof result.value !== "string") {
      return "";
    }

    const value = result.value;
    this.emitAccess({ type: KeyType.Image, scope: result.scope, key, value });
    return value;
  }

  public imageList(key: string): readonly string[] {
    const result = this.findRecordValue((scope) => scope.images, key);
    if (result === undefined) {
      return EMPTY_IMAGE_LIST;
    }

    const value = typeof result.value === "string"
      ? Object.freeze([result.value])
      : Object.freeze([...result.value]);

    this.emitAccess({
      type: KeyType.ImageList,
      scope: result.scope,
      key,
      value,
    });
    return value;
  }

  public route(key: number): string {
    const result = this.findRoute(key);
    if (result === undefined) {
      return "";
    }

    this.emitAccess({ type: KeyType.Route, ...result, key });
    return result.value;
  }

  public size(key: string): number {
    const result = this.findRecordValue((scope) => scope.sizes, key);
    if (result === undefined) {
      return 14;
    }

    this.emitAccess({ type: KeyType.Size, ...result, key });
    return result.value;
  }

  public padding(key: string): number {
    const result = this.findRecordValue((scope) => scope.paddings, key);
    if (result === undefined) {
      return 0;
    }

    this.emitAccess({ type: KeyType.Padding, ...result, key });
    return result.value;
  }

  public margin(key: string): number {
    const result = this.findRecordValue((scope) => scope.margins, key);
    if (result === undefined) {
      return 0;
    }

    this.emitAccess({ type: KeyType.Margin, ...result, key });
    return result.value;
  }

  public misc(key: string): unknown | null {
    const result = this.findRecordValue((scope) => scope.misc, key);
    if (result === undefined) {
      return null;
    }

    this.emitAccess({ type: KeyType.Misc, ...result, key });
    return result.value;
  }

  public textStyle(key: string): ConfigTextStyle {
    const result = this.findRecordValue((scope) => scope.textStyles, key);
    if (result === undefined) {
      return EMPTY_TEXT_STYLE;
    }

    this.emitAccess({ type: KeyType.TextStyle, ...result, key });
    return result.value;
  }

  /**
   * Return the full translation map from the winning scope that contains key.
   * This intentionally retains the Dart runtime's currentTranslations contract.
   */
  public translations(key: string): ConfigTranslations {
    const result = this.findRecordValue((scope) => scope.translations, key);
    if (result === undefined) {
      return EMPTY_TRANSLATIONS;
    }

    const value = result.scope.translations;
    this.emitAccess({
      type: KeyType.Translation,
      scope: result.scope,
      key,
      value,
    });
    return value;
  }

  /** Dart-compatible alias for {@link translations}. */
  public currentTranslations(key: string): ConfigTranslations {
    return this.translations(key);
  }

  /** Clear all subscriptions held by this configuration instance. */
  public dispose(): void {
    this.changeListeners.clear();
    this.accessListeners.clear();
  }

  private findRecordValue<T>(
    select: (scope: ConfigScope) => Readonly<Record<string, T>>,
    key: string,
  ): LookupResult<T> | undefined {
    let winner: LookupResult<T> | undefined;

    for (const scope of this.scopesValue) {
      const values = select(scope);
      if (!Object.hasOwn(values, key)) {
        continue;
      }

      if (winner === undefined || scope.weight >= winner.scope.weight) {
        winner = { scope, value: values[key]! };
      }
    }

    return winner;
  }

  private findRoute(key: number): LookupResult<string> | undefined {
    let winner: LookupResult<string> | undefined;

    for (const scope of this.scopesValue) {
      if (!scope.routes.has(key)) {
        continue;
      }

      if (winner === undefined || scope.weight >= winner.scope.weight) {
        winner = { scope, value: scope.routes.get(key)! };
      }
    }

    return winner;
  }

  private emitChange(notify: boolean): void {
    if (!notify) {
      return;
    }

    for (const listener of [...this.changeListeners]) {
      listener(this);
    }
  }

  private emitAccess(event: ConfigAccessEvent): void {
    const immutableEvent = Object.freeze(event);
    for (const listener of [...this.accessListeners]) {
      listener(immutableEvent);
    }
  }
}

File: configurator_python/src/configurator/configuration.py
"""Weighted configuration resolution and event subscriptions."""

from __future__ import annotations

from collections.abc import Callable, Iterable, Mapping, Sequence
from itertools import count
from threading import RLock
from time import time
from typing import Any, TypeAlias, TypeVar, cast

from .models import ConfigAccessLog, ConfigScope, KeyType, TextStyle, TranslationMap


Unsubscribe: TypeAlias = Callable[[], None]
ChangeCallback: TypeAlias = Callable[["Configuration"], None]
AccessCallback: TypeAlias = Callable[[ConfigAccessLog[Any, Any]], None]

K = TypeVar("K")
V = TypeVar("V")
CallbackT = TypeVar("CallbackT", ChangeCallback, AccessCallback)


class Configuration:
    """Resolve configuration values from a weighted stack of scopes.

    A higher weight wins. Scope insertion order breaks weight ties, with the
    later scope taking precedence. The configuration always retains at least
    one scope so `current_scope_name` remains available.
    """

    def __init__(self, scopes: Iterable[ConfigScope] = ()) -> None:
        initial = list(scopes)
        self._scopes: list[ConfigScope] = initial or [self._default_scope()]
        self._lock = RLock()
        self._change_callbacks: dict[int, ChangeCallback] = {}
        self._access_callbacks: dict[int, AccessCallback] = {}
        self._subscription_ids = count()

    @staticmethod
    def _default_scope() -> ConfigScope:
        return ConfigScope.empty(str(int(time() * 1000)))

    @property
    def scopes(self) -> tuple[ConfigScope, ...]:
        """A snapshot of scopes in insertion order."""

        with self._lock:
            return tuple(self._scopes)

    @property
    def current_scope_name(self) -> str:
        """The name of the most recently inserted scope."""

        with self._lock:
            return self._scopes[-1].name

    def push_scope(
        self,
        scope: ConfigScope,
        *,
        check_equality: bool = True,
        notify: bool = True,
    ) -> bool:
        """Append a scope, returning whether the configuration changed."""

        with self._lock:
            if check_equality and scope == self._scopes[-1]:
                return False
            self._scopes.append(scope)

        if notify:
            self.notify_listeners()
        return True

    def pop_scope(self, *, notify: bool = True) -> ConfigScope | None:
        """Remove and return the latest scope, preserving the final scope."""

        with self._lock:
            if len(self._scopes) == 1:
                return None
            removed = self._scopes.pop()

        if notify:
            self.notify_listeners()
        return removed

    def pop_scope_until(
        self,
        predicate: Callable[[ConfigScope], bool],
        *,
        notify: bool = True,
    ) -> tuple[ConfigScope, ...]:
        """Pop scopes until `predicate` accepts the current scope."""

        removed: list[ConfigScope] = []
        with self._lock:
            while len(self._scopes) > 1 and not predicate(self._scopes[-1]):
                removed.append(self._scopes.pop())

        if removed and notify:
            self.notify_listeners()
        return tuple(removed)

    def remove_scope_where(
        self,
        predicate: Callable[[ConfigScope], bool],
        *,
        notify: bool = True,
    ) -> tuple[ConfigScope, ...]:
        """Remove all matching scopes and return them in insertion order."""

        with self._lock:
            matches = [(scope, predicate(scope)) for scope in self._scopes]
            removed = [scope for scope, should_remove in matches if should_remove]
            if not removed:
                return ()
            self._scopes = [
                scope for scope, should_remove in matches if not should_remove
            ]
            if not self._scopes:
                self._scopes.append(self._default_scope())

        if notify:
            self.notify_listeners()
        return tuple(removed)

    def remove_last_scope_where(
        self,
        predicate: Callable[[ConfigScope], bool],
        *,
        notify: bool = True,
    ) -> ConfigScope | None:
        """Remove and return the latest scope accepted by `predicate`."""

        removed: ConfigScope | None = None
        with self._lock:
            for index in range(len(self._scopes) - 1, -1, -1):
                if predicate(self._scopes[index]):
                    removed = self._scopes.pop(index)
                    break
            if removed is None:
                return None
            if not self._scopes:
                self._scopes.append(self._default_scope())

        if notify:
            self.notify_listeners()
        return removed

    def subscribe(
        self,
        callback: ChangeCallback,
        *,
        emit_current: bool = False,
    ) -> Unsubscribe:
        """Subscribe to mutations and return an idempotent unsubscribe callback."""

        subscription_id = self._add_subscription(self._change_callbacks, callback)
        if emit_current:
            callback(self)
        return self._unsubscribe_callback(self._change_callbacks, subscription_id)

    def subscribe_access(self, callback: AccessCallback) -> Unsubscribe:
        """Subscribe to successful value lookups."""

        subscription_id = self._add_subscription(self._access_callbacks, callback)
        return self._unsubscribe_callback(self._access_callbacks, subscription_id)

    def notify_listeners(self) -> None:
        """Notify current change subscribers with this configuration."""

        with self._lock:
            callbacks = tuple(self._change_callbacks.values())
        for callback in callbacks:
            callback(self)

    def flag(self, key: str) -> bool:
        resolved = self._resolve("flags", key)
        if resolved is None:
            return False
        scope, raw = resolved
        value = raw is True
        self._emit_access(KeyType.FLAG, scope, key, value)
        return value

    def color(self, key: str) -> str:
        resolved = self._resolve("colors", key)
        if resolved is None:
            return ""
        scope, value = resolved
        self._emit_access(KeyType.COLOR, scope, key, value)
        return cast(str, value)

    def image(self, key: str) -> str:
        resolved = self._resolve("images", key)
        if resolved is None:
            return ""
        scope, raw = resolved
        value = raw if isinstance(raw, str) else ""
        self._emit_access(KeyType.IMAGE, scope, key, value)
        return value

    def image_list(self, key: str) -> list[str]:
        resolved = self._resolve("images", key)
        if resolved is None:
            return []
        scope, raw = resolved
        if isinstance(raw, str):
            value = [raw]
        elif isinstance(raw, Sequence):
            value = [item for item in raw if isinstance(item, str)]
        else:
            value = []
        self._emit_access(KeyType.IMAGE_LIST, scope, key, value)
        return value

    def route(self, key: int) -> str:
        resolved = self._resolve("routes", key)
        if resolved is None:
            return ""
        scope, raw = resolved
        value = raw if isinstance(raw, str) else ""
        self._emit_access(KeyType.ROUTE, scope, key, value)
        return value

    def size(self, key: str) -> float:
        return self._number("sizes", KeyType.SIZE, key, 14.0)

    def padding(self, key: str) -> float:
        return self._number("paddings", KeyType.PADDING, key, 0.0)

    def margin(self, key: str) -> float:
        return self._number("margins", KeyType.MARGIN, key, 0.0)

    def misc(self, key: str) -> Any:
        resolved = self._resolve("misc", key)
        if resolved is None:
            return None
        scope, value = resolved
        self._emit_access(KeyType.MISC, scope, key, value)
        return value

    def text_style(self, key: str) -> TextStyle:
        resolved = self._resolve("text_styles", key)
        if resolved is None:
            return {}
        scope, value = resolved
        self._emit_access(KeyType.TEXT_STYLE, scope, key, value)
        return cast(TextStyle, value)

    def translations(self, key: str) -> TranslationMap:
        resolved = self._resolve("translations", key)
        if resolved is None:
            return {}
        scope, _ = resolved
        value = scope.translations
        self._emit_access(KeyType.TRANSLATION, scope, key, value)
        return value

    def current_translations(self, key: str) -> TranslationMap:
        """Dart-compatible alias for `translations`."""

        return self.translations(key)

    def _number(
        self,
        mapping_name: str,
        key_type: KeyType,
        key: str,
        default: float,
    ) -> float:
        resolved = self._resolve(mapping_name, key)
        if resolved is None:
            return default
        scope, raw = resolved
        value = float(cast(float, raw))
        self._emit_access(key_type, scope, key, value)
        return value

    def _resolve(self, mapping_name: str, key: K) -> tuple[ConfigScope, Any] | None:
        with self._lock:
            winner: tuple[int, int, ConfigScope, Mapping[Any, Any]] | None = None
            for index, scope in enumerate(self._scopes):
                values = cast(Mapping[Any, Any], getattr(scope, mapping_name))
                if key not in values:
                    continue
                candidate = (scope.weight, index, scope, values)
                if winner is None or candidate[:2] > winner[:2]:
                    winner = candidate

            if winner is None:
                return None
            return winner[2], winner[3][key]

    def _emit_access(
        self,
        key_type: KeyType,
        scope: ConfigScope,
        key: K,
        value: V,
    ) -> None:
        event = ConfigAccessLog(key_type, scope, key, value)
        with self._lock:
            callbacks = tuple(self._access_callbacks.values())
        for callback in callbacks:
            callback(event)

    def _add_subscription(
        self,
        subscriptions: dict[int, CallbackT],
        callback: CallbackT,
    ) -> int:
        with self._lock:
            subscription_id = next(self._subscription_ids)
            subscriptions[subscription_id] = callback
        return subscription_id

    def _unsubscribe_callback(
        self,
        subscriptions: dict[int, Any],
        subscription_id: int,
    ) -> Unsubscribe:
        def unsubscribe() -> None:
            with self._lock:
                subscriptions.pop(subscription_id, None)

        return unsubscribe
